src/core/io/device/io_controller.cpp

```diff
@@ -66,14 +66,47 @@ void IODeviceController::addHandle(HandleDescriptor &paHandleDescriptor) {
                    paHandleDescriptor.mId.c_str());
     return;
   }
+  addHandle(paHandleDescriptor.mId, std::move(handle));
+}
 
-  switch (handle->getDirection()) {
-    case IOMapper::In: addHandle(mInputHandles, paHandleDescriptor.mId, std::move(handle)); break;
-    case IOMapper::Out: addHandle(mOutputHandles, paHandleDescriptor.mId, std::move(handle)); break;
+void IODeviceController::addHandle(std::string const &paId, std::unique_ptr<IOHandle> paHandle) {
+  switch (paHandle->getDirection()) {
+    case IOMapper::In: addHandle(mInputHandles, paId, std::move(paHandle)); break;
+    case IOMapper::Out: addHandle(mOutputHandles, paId, std::move(paHandle)); break;
+    case IOMapper::InOut: addHandle(mDiverseHandles, paId, std::move(paHandle)); break;
+    case IOMapper::UnknownDirection: addHandle(mDiverseHandles, paId, std::move(paHandle)); break;
     default: break;
   }
 }
 
+void IODeviceController::updateHandleList(std::string const &paId, IOHandle *paHandle) {
+  for (auto it = mDiverseHandles.begin(); it != mDiverseHandles.end(); ++it) {
+    if (it->get() == paHandle) {
+      auto handleDirection = (*it)->getDirection();
+      switch (handleDirection) {
+        case IOMapper::In: {
+          CCriticalRegion criticalRegion(mHandleMutex);
+          mInputHandles.push_back(std::move(*it));
+          mDiverseHandles.erase(it);
```

Comment:
this will invalidate the iterator. you need to use the return value of erease as the new iterator. But then you can not use ++it in the for loop